txp/Pos/Txp/Toil/Monad.hs

```diff
@@ -0,0 +1,274 @@
+    deriving (Functor)
+
+-- | Mutable state used in global Toil.
+data GlobalToilState = GlobalToilState
+    { _gtsUtxoModifier :: !UtxoModifier
+    , _gtsStakesView   :: !StakesView
+    }
+
+-- | Default 'GlobalToilState'.
+defGlobalToilState :: GlobalToilState
+defGlobalToilState =
+    GlobalToilState {_gtsUtxoModifier = mempty, _gtsStakesView = def}
+
+makeLenses ''GlobalToilState
+
+-- | Immutable environment used in global Toil.
+data GlobalToilEnv = GlobalToilEnv
+    { _gteUtxo       :: !UtxoLookup
+    , _gteTotalStake :: !Coin
+    }
+
+makeLenses ''GlobalToilEnv
+
+-- | Base monad in which global Toil happens.
+type GlobalToilMBase = NamedPureLogger (F StakesLookupF)
+
+-- | Monad in which global Toil happens.
+type GlobalToilM
+     = ReaderT GlobalToilEnv (StateT GlobalToilState GlobalToilMBase)
+
+-- | Run given action in some monad capable of getting stakeholders'
+-- stakes and logging.
+runGlobalToilMBase ::
+       forall m a. (WithLogger m)
+    => (StakeholderId -> m (Maybe Coin))
+    -> GlobalToilMBase a
+    -> m a
+runGlobalToilMBase stakeGetter = launchNamedPureLog foldF'
+  where
+    foldF' :: forall x. F StakesLookupF x -> m x
+    foldF' =
+        foldF $ \case
+            StakesLookupF sId f -> f <$> stakeGetter sId
+
+-- | Run 'GlobalToilM' action in some monad capable of getting
+-- stakeholders' stakes and logging.
+runGlobalToilM ::
+       forall m a. (WithLogger m)
+    => GlobalToilEnv
+    -> GlobalToilState
+    -> (StakeholderId -> m (Maybe Coin))
+    -> GlobalToilM a
+    -> m (a, GlobalToilState)
+runGlobalToilM env gts stakeGetter =
+    runGlobalToilMBase stakeGetter . usingStateT gts . usingReaderT env
+
+-- | Get stake of a given stakeholder.
+getStake :: StakeholderId -> GlobalToilM (Maybe Coin)
+getStake id =
+    (<|>) <$> use (gtsStakesView . svStakes . at id) <*> baseLookup id
+  where
+    baseLookup :: StakeholderId -> GlobalToilM (Maybe Coin)
+    baseLookup i =
+        lift $ lift $ lift $ F $ \kPure kFree -> kFree (StakesLookupF i kPure)
+
+-- | Get total stake of all stakeholders.
+getTotalStake :: GlobalToilM Coin
+getTotalStake =
+    maybe (view gteTotalStake) pure =<< use (gtsStakesView . svTotal)
+
+-- | Set stake of a given stakeholder.
+setStake :: StakeholderId -> Coin -> GlobalToilM ()
+setStake id c = gtsStakesView . svStakes . at id .= Just c
+
+-- | Set total stake of all stakeholders.
+setTotalStake :: Coin -> GlobalToilM ()
+setTotalStake c = gtsStakesView . svTotal .= Just c
+
+-- | Extended version of 'GlobalToilM'. It allows to put extra data
+-- into reader context and extra state. It's needed for explorer which
+-- has more complicated transaction processing.
+type ExtendedGlobalToilM extraEnv extraState =
+    ReaderT (GlobalToilEnv, extraEnv) (
+        StateT (GlobalToilState, extraState) (
+            GlobalToilMBase
+    ))
+
+-- | Natural transformation from 'GlobalToilM to 'ExtendedGlobalToilM'.
+natGlobalToilM :: GlobalToilM ~> ExtendedGlobalToilM __ ___
```

Comment:
use `extraEnv` and `extraState` instead of `__` and `___` please, and let's rename it to `extendGlobalToilM`